src/menu/Menu.tsx

```diff
@@ -0,0 +1,274 @@
+		onValue,
+		onRequestClose,
+		onActiveIndexChange,
+		activeIndex,
+		focusable = true,
+		onBlur,
+		onFocus,
+		numberInView,
+		itemRenderer
+	} = properties();
+
+	if (initialValue !== undefined && initialValue !== icache.get('initial')) {
+		icache.set('initial', initialValue);
+		icache.set('value', initialValue);
+		icache.set('activeIndex', findIndex(options, (option) => option.value === initialValue));
+	}
+
+	if (numberInView && numberInView !== icache.get('numberInView')) {
+		icache.set('numberInView', numberInView);
+
+		const offscreenItemProps = {
+			selected: false,
+			onSelect: () => {},
+			active: false,
+			onRequestActive: () => {},
+			onActive: () => {},
+			scrollIntoView: false
+		};
+
+		const itemHeight = icache.getOrSet(
+			'itemHeight',
+			offscreenHeight(
+				<MenuItem {...offscreenItemProps}>
+					{itemRenderer
+						? itemRenderer({
+								selected: false,
+								active: false,
+								value: 'offscreen',
+								disabled: false
+						  })
+						: 'offscreen'}
+				</MenuItem>
+			)
+		);
+		itemHeight && icache.set('menuHeight', numberInView * itemHeight);
+	}
+
+	const selectedValue = icache.get('value');
+	const computedActiveIndex =
+		activeIndex === undefined ? icache.getOrSet('activeIndex', 0) : activeIndex;
+
+	function _setActiveIndex(index: number) {
+		if (onActiveIndexChange) {
+			onActiveIndexChange(index);
+		} else {
+			icache.set('activeIndex', index);
+		}
+	}
+
+	function _setValue(value: string) {
+		icache.set('value', value);
+		onValue(value);
+	}
+
+	function _onKeyDown(event: KeyboardEvent) {
+		event.stopPropagation();
+
+		switch (event.which) {
+			case Keys.Enter:
+			case Keys.Space:
+				event.preventDefault();
+				const activeItem = options[computedActiveIndex];
+				!activeItem.disabled && _setValue(activeItem.value);
+				break;
+			case Keys.Down:
+				event.preventDefault();
+				_setActiveIndex((computedActiveIndex + 1) % options.length);
+				break;
+			case Keys.Up:
+				event.preventDefault();
+				_setActiveIndex((computedActiveIndex - 1 + options.length) % options.length);
+				break;
+			case Keys.Escape:
+				event.preventDefault();
+				onRequestClose && onRequestClose();
+				break;
+			default:
+				const newIndex = getComputedIndexFromInput(event.key);
+				if (newIndex !== undefined) {
+					_setActiveIndex(newIndex);
+				}
+		}
+	}
+
+	function getComputedIndexFromInput(key: string) {
+		const existingTimer = cache.get<NodeJS.Timer>('resetInputTextTimer');
+		let inputText = cache.get<string>('inputText') || '';
+		existingTimer && clearTimeout(existingTimer);
+
+		cache.set(
+			'resetInputTextTimer',
+			setTimeout(() => {
+				cache.set('inputText', '');
+			}, 800)
+		);
+
+		inputText += `${key}`;
+		cache.set('inputText', inputText);
+
+		return findIndex(options, ({ disabled, value, label }, i) => {
+			if (disabled) {
+				return false;
+			}
+
+			if ((label || value).toLowerCase().indexOf(inputText.toLowerCase()) === 0) {
```

Comment:
Should the user be able to pass in their own comparison function to do case-sensitive matching?

Reply:
again, this is a relatively opinionated widget that deals with text labels / values. this is simply allowing the user to type `n` when the menu is focused to select `nevada` (in the states example)